TELCODOCS-1506 Telco CORE Reference Design Specification (#65732)
openshift/openshift-docs: modules/telco-core-cpu-partitioning-performance-tune.adoc

+
For more information, see https://docs.openshift.com/container-platform/latest/scalability_and_performance/cnf-low-latency-tuning.html#cnf-cpu-infra-container_cnf-master[CNF Low Latency Tuning]
+
Configure system level performance.
For recommended settings, see link:https://docs.openshift.com/container-platform/latest/scalability_and_performance/ztp_far_edge/ztp-reference-cluster-configuration-for-vdu.html#ztp-du-configuring-host-firmware-requirements_sno-configure-for-vdu[Configuring host firmware for low latency and high performance].

Limits and requirements::
* The operating system needs a certain amount of CPU to perform all the support tasks including kernel networking.
** A system with just user plane networking applications (DPDK) needs at least one Core (2 hyperthreads when enabled) reserved for the operating system and the infrastructure components.

> * A system with Hyper-Threading enabled must always put all core sibling threads to the same pool of CPUs.

Reviewer: Not sure if there is a difference, but you have "hyperthread" and "hyper-thread" both in these docs. May want to use one or the other for consistency.

Author: Looks like hyperthreading is used in docs funny I thought I got a different recommendation

Author: Though I got vale error to us Hyper-Threading plus looks like we use both in our docs. I guess I should go with recommendation from vale as someone wrote that rule.

> * The set of reserved + isolated cores must include all CPU cores.

Reviewer: ```suggestion
* The set of reserved and isolated cores must include all CPU cores.
```

* Core 0 of each NUMA node must be included in the reserved CPU set.

> * Isolated cores might still be impacted by interrupts. The following annotations must be attached to the Pod if guaranteed QoS pods require full use of CPU:

Reviewer: > 
```suggestion
* Isolated cores might be impacted by interrupts. The following annotations must be attached to the pod if guaranteed QoS pods require full use of the CPU:
```

+

> ----

Reviewer: I think a blank line is needed above the code block per [guidelines](https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#empty-line-before-code-blocks).

Author: seems to mess up the indentation and pushes that block out.

cpu-load-balancing.crio.io: "disable"
cpu-quota.crio.io: "disable"
irq-load-balancing.crio.io: "disable"
----

> * When per pod power management is enabled (`PerformanceProfile.workloadHints.perPodPowerManagement`) the following annotations must also be attached to the pod if guaranteed QoS pods require full use of CPU:

Reviewer: ```suggestion
* When per-pod power management is enabled with `PerformanceProfile.workloadHints.perPodPowerManagement`, the following annotations must also be attached to the pod if guaranteed QoS pods require full use of the CPU:
```

> +

Reviewer: see previous comment about blank line above code block

----
cpu-c-states.crio.io: "disable"
cpu-freq-governor.crio.io: "performance"
----

Engineering considerations::

> * The minimum reserved capacity (`systemReserved`) required can be found according to the OpenShift Container Platform solution guide link:https://access.redhat.com/solutions/5843241[Which amount of CPU and memory are recommended to reserve for the system in OCP 4 nodes?]

Reviewer: If this is considered a "guide" then I think it needs italics for the title. If it's just an article, then I think quotation marks?

> * The actual required reserved CPU capacity is further dependent on the cluster configuration and workload attributes.

Reviewer: ```suggestion
* The actual required reserved CPU capacity depends on the cluster configuration and workload attributes.
```

* This reserved CPU value must be rounded up to a full core (2 hyper-thread) alignment.
* Changes to the CPU partitioning will drain and reboot the nodes in the MCP.

> * The reserved CPUs reduce the Pod density as the reserved CPUs are removed from the Allocatable capacity of the OpenShift node.

Reviewer: ```suggestion
* The reserved CPUs reduce the pod density, as the reserved CPUs are removed from the allocatable capacity of the OpenShift node.
```

> * The realtime workload hint should be enabled if the workload is real-time capable.

Reviewer: ```suggestion
* The real-time workload hint should be enabled if the workload is real-time capable.
```

> * Hardware without IRQ affinity support will impact isolated CPUs. To ensure that pods with guaranteed cpu QoS have full use of allocated CPU, all hardware in the server must support IRQ affinity.

Reviewer: - spell out IRQ at first use?
- CPU should be caps
